### PR TITLE
analyze: analyze table after restore

### LINKED ISSUE #598: Analyze database after restore
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

After restoring a database, TiDB may choose a wrong plan for selects.

2. What did you expect to see?

After restoring a database, BR should issue an analysis automatically or restore analysis data too.

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve? 
close #598 

### What is changed and how it works?
1. backup stats json in backup meta.
2. restore stats after validate checksum.

### Check List 

Tests 
 - Manual test (add detailed scripts or steps below)

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Release Note

 - Fix the issue that didn't analyze table after restoration.

### CODE REVIEW

File: pkg/backup/client.go [hunk 1]
```diff
@@ -331,9 +333,20 @@ func BuildBackupRangeAndSchema(
 				return nil, nil, errors.Trace(err)
 			}
 
+			jsonTable, err := h.DumpStatsToJSON(dbInfo.Name.String(), tableInfo, nil)
+			if err != nil {
+				return nil, nil, errors.Trace(err)
+			}
+			log.Info("JsonTable:", zap.Any("t", jsonTable))
```
Comment (kennytm) [inline pkg/backup/client.go:340]:
```suggestion
```

File: cmd/backup.go [hunk 2]
```diff
@@ -56,8 +55,6 @@ func NewBackupCommand() *cobra.Command {
 
 			// Do not run ddl worker in BR.
 			ddl.RunWorker = false
-			// Do not run stat worker in BR.
-			session.DisableStats4Test()
```
Comment (kennytm) [inline cmd/backup.go:60]: can we disable the stats loop after we have dumped all JSONTables?

Comment (3pointer) [reply cmd/backup.go:60]: when we reached here, we have domain already. so it seems we can not touch the `statsLease`, https://github.com/pingcap/tidb/blob/e136429d8dc5d70f43cd3f94179b0b9f47595097/session/tidb.go#L68

**Review by kennytm — COMMENTED** [review]
please add a test involving restoring from an older archive where the stats are missing (nil).

(perhaps add a flag like `br backup --stats=false` too to simplify testing)

File: pkg/restore/batcher_test.go [hunk 3]
```diff
@@ -68,52 +68,59 @@ func newDrySender() *drySender {
...
 	}
 	return nil
 }
 
 func (manager recordCurrentTableManager) Leave(_ context.Context, tables []restore.CreatedTable) error {
 	for _, t := range tables {
-		if !manager[t.Table.ID] {
+		manager.Lock()
+		if !manager.m[t.Table.ID] {
 			return errors.Errorf("Table %d is removed before added", t.Table.ID)
```
Comment (lichunzhu) [inline pkg/restore/batcher_test.go:101]: Should we unlock `manager` here?

File: pkg/restore/batcher_test.go [hunk 4]
```diff
@@ -68,52 +68,62 @@ func newDrySender() *drySender {
...
+		m: make(map[int64]bool),
+	}
 }
 
-func (manager recordCurrentTableManager) Enter(_ context.Context, tables []restore.CreatedTable) error {
+func (manager *recordCurrentTableManager) Enter(_ context.Context, tables []restore.CreatedTable) error {
 	for _, t := range tables {
 		log.Info("entering", zap.Int64("table ID", t.Table.ID))
-		manager[t.Table.ID] = true
+		manager.lock.Lock()
```
Comment (glorv) [inline pkg/restore/batcher_test.go:92]: move this lock outsize the for loop

File: pkg/restore/batcher_test.go [hunk 5]
```diff
@@ -68,52 +68,62 @@ func newDrySender() *drySender {
...
+		manager.lock.Unlock()
 	}
 	return nil
 }
 
-func (manager recordCurrentTableManager) Leave(_ context.Context, tables []restore.CreatedTable) error {
+func (manager *recordCurrentTableManager) Leave(_ context.Context, tables []restore.CreatedTable) error {
 	for _, t := range tables {
-		if !manager[t.Table.ID] {
+		manager.lock.Lock()
```
Comment (glorv) [inline pkg/restore/batcher_test.go:101]: ditto

File: pkg/restore/batcher_test.go [hunk 6]
```diff
@@ -68,52 +68,62 @@ func newDrySender() *drySender {
...
-func (manager recordCurrentTableManager) Has(tables ...restore.TableWithRange) bool {
+func (manager *recordCurrentTableManager) Has(tables ...restore.TableWithRange) bool {
 	ids := make([]int64, 0, len(tables))
-	currentIDs := make([]int64, 0, len(manager))
+	currentIDs := make([]int64, 0, len(manager.m))
 	for _, t := range tables {
 		ids = append(ids, t.Table.ID)
 	}
-	for id, contains := range manager {
+	for id, contains := range manager.m {
```
Comment (glorv) [inline pkg/restore/batcher_test.go:119]: Read option should be protected by lock

**Comment by 3pointer** [comment]
> please add a test involving restoring from an older archive where the stats are missing (nil).
> 
> (perhaps add a flag like `br backup --stats=false` too to simplify testing)

Done， PTAL

File: pkg/task/backup.go [hunk 7]
```diff
@@ -119,8 +126,11 @@ func (cfg *BackupConfig) ParseFromFlags(flags *pflag.FlagSet) error {
 	if err = cfg.Config.ParseFromFlags(flags); err != nil {
 		return errors.Trace(err)
 	}
-
 	cfg.RemoveSchedulers, err = flags.GetBool(flagRemoveSchedulers)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cfg.IgnoreStats, err = flags.GetBool(flagIgnoreStats)
 	return err
```
Comment (lichunzhu) [inline pkg/task/backup.go:134]: Why don't we `errors.Trace` this `err`?

Comment (3pointer) [reply pkg/task/backup.go]: Fixed

**Review by lichunzhu — COMMENTED** [review]
LGTM

File: pkg/restore/batcher_test.go [hunk 8]
```diff
@@ -68,52 +68,63 @@ func newDrySender() *drySender {
...
-type recordCurrentTableManager map[int64]bool
+type recordCurrentTableManager struct {
+	lock sync.Mutex
+	m    map[int64]bool
+}
 
-func (manager recordCurrentTableManager) Close(ctx context.Context) {
-	if len(manager) > 0 {
+func (manager *recordCurrentTableManager) Close(ctx context.Context) {
+	if len(manager.m) > 0 {
```
Comment (kennytm) [inline pkg/restore/batcher_test.go:77]: it is not thread-safe to read the map size outside of the lock

File: pkg/restore/batcher_test.go [hunk 9]
```diff
@@ -68,52 +68,63 @@ func newDrySender() *drySender {
...
+		delete(manager.m, t.Table.ID)
 	}
 	return nil
 }
 
-func (manager recordCurrentTableManager) Has(tables ...restore.TableWithRange) bool {
+func (manager *recordCurrentTableManager) Has(tables ...restore.TableWithRange) bool {
 	ids := make([]int64, 0, len(tables))
-	currentIDs := make([]int64, 0, len(manager))
+	currentIDs := make([]int64, 0, len(manager.m))
```
Comment (kennytm) [inline pkg/restore/batcher_test.go:114]: ditto

File: pkg/restore/client.go [hunk 10]
```diff
@@ -775,6 +794,17 @@ func (rc *Client) execChecksum(ctx context.Context, tbl CreatedTable, kvClient k
...
 	}
+	if table.Stats != nil {
+		log.Info("start loads analyze after validate checksum",
+			zap.Stringer("db name", tbl.OldTable.DB.Name),
+			zap.Stringer("name", tbl.OldTable.Info.Name),
+			zap.Int64("old id", tbl.OldTable.Info.ID),
+			zap.Int64("new id", tbl.Table.ID),
+		)
+		if err := rc.statsHandler.LoadStatsFromJSON(rc.dom.InfoSchema(), table.Stats); err != nil {
+			log.Error("analyze table failed, ", zap.Any("table", table.Stats), zap.Error(err))
```
Comment (kennytm) [inline pkg/restore/client.go:805]:
```suggestion
			log.Error("analyze table failed", zap.Any("table", table.Stats), zap.Error(err))
```

**Review by kennytm — COMMENTED** [review]
rest LGTM

**Comment by kennytm** [comment]
LGTM

**Review by ti-srebot — APPROVED** [review]
